lib/bundler/vendored_persistent.rb

```diff
@@ -15,3 +15,31 @@ module HTTP
   end
 end
 require "bundler/vendor/net-http-persistent/lib/net/http/persistent"
+
+module Bundler
+  class PersistentHTTP < Persistent::Net::HTTP::Persistent
+    def connection_for(uri)
+      connection = super
+      warn_old_tls_version_rubygems_connection(uri, connection)
+      connection
+    end
+
+    def warn_old_tls_version_rubygems_connection(uri, connection)
+      return unless connection.use_ssl?
+      return unless (uri.hostname || "").end_with?("rubygems.org")
+
+      socket = connection.instance_variable_get(:@socket)
+      ssl_version = socket.io.ssl_version
+
+      case ssl_version
+      when /TLSv([\d\.]+)/
+        version = Gem::Version.new($1)
+        if version < Gem::Version.new("1.1")
+          Bundler.ui.warn "Your Ruby version does not support TLSv1.1 or newer" \
+            ", which will be required to connect to https://#{uri.hostname}" \
+            " by June 2018."
```

Comment:
I'd like to say January 2018. That gives us a buffer and we can see who still can't upgrade.